pkg/controller/daemon/update.go

```diff
@@ -294,24 +295,44 @@ func (dsc *DaemonSetsController) controlledHistories(ds *extensions.DaemonSet) (
 	return result, nil
 }
 
-// Match check if ds template is semantically equal to the template stored in history
-func Match(template *v1.PodTemplateSpec, history *apps.ControllerRevision) (bool, error) {
-	t, err := DecodeHistory(history)
-	return apiequality.Semantic.DeepEqual(template, t), err
-}
-
-func DecodeHistory(history *apps.ControllerRevision) (*v1.PodTemplateSpec, error) {
-	template := v1.PodTemplateSpec{}
-	err := json.Unmarshal(history.Data.Raw, &template)
-	return &template, err
+// Match check if the given DaemonSet's template matches the template stored in the given history.
+func Match(ds *extensions.DaemonSet, history *apps.ControllerRevision) (bool, error) {
+	patch, err := getPatch(ds)
+	if err != nil {
+		return false, err
+	}
+	return bytes.Equal(patch, history.Data.Raw), nil
 }
 
-func encodeTemplate(template *v1.PodTemplateSpec) ([]byte, error) {
-	return json.Marshal(template)
+// getPatch returns a strategic merge patch that can be applied to restore a Daemonset to a
+// previous version. If the returned error is nil the patch is valid. The current state that we save is just the
+// PodSpecTemplate. We can modify this later to encompass more state (or less) and remain compatible with previously
+// recorded patches.
+func getPatch(ds *extensions.DaemonSet) ([]byte, error) {
+	dsBytes, err := json.Marshal(ds)
+	if err != nil {
+		return nil, err
+	}
+	var raw map[string]interface{}
+	err = json.Unmarshal(dsBytes, &raw)
+	if err != nil {
+		return nil, err
+	}
+	objCopy := make(map[string]interface{})
+	specCopy := make(map[string]interface{})
+
+	// Create a patch of the DaemonSet that replaces spec.template
+	spec := raw["spec"].(map[string]interface{})
+	template := spec["template"].(map[string]interface{})
+	specCopy["template"] = template
+	template["$patch"] = "replace"
+	objCopy["spec"] = specCopy
```

Comment:
For DaemonSet history we only care about template changes, and when rolling back, we only roll back templates